File: tools/playtest_break_scenarios/driving_physics.py
"""Driving/physics abuse: floor it, reverse it, coast it, dynamite it.

Every scenario here puts the truck in a position no sane driver would choose
and checks whether the physics, the ledgers, and the spoken cues stay honest
about the consequences.
"""

from __future__ import annotations

import re

from playtest_break import DT, Rig, _fabricated_curve, _outcome, scenario


@scenario(
    "floor_it_through_town",
    "Hold the floor through urban speed zones; strikes and spoken fine totals must agree.",
)
def _floor_it():
    rig = Rig()
    findings: list[str] = []
    try:
        d = rig.d
        rig.prepare(speed_mph=30.0)
        rig.held.add(rig.pygame.K_UP)
        rig.step(9000, until=lambda: d.trip.position_mi >= min(25.0, d.trip.total_miles - 8.0))
        strike_lines = rig.lines_with("Speeding strike")
        if d.speeding_strikes != len(strike_lines):
            findings.append(
                f"strike ledger mismatch: {d.speeding_strikes} strikes recorded, "
                f"{len(strike_lines)} spoken"
            )
        from freight_fate.states.driving_core import _speeding_settlement_fine

        totals = []
        for line in strike_lines:
            m = re.search(r"total ([\d,]+) dollars|the ([\d,]+)-dollar maximum", line)
            if m:
                totals.append(float((m.group(1) or m.group(2)).replace(",", "")))
        for n, spoken_total in enumerate(totals, start=1):
            expected = _speeding_settlement_fine(n)
            if abs(spoken_total - expected) > 0.01:
                findings.append(
                    f"spoken fine total {spoken_total:,.0f} != ledger {expected:,.0f} "
                    f"after strike {n}"
                )
        if d.speeding_strikes == 0:
            findings.append("held the floor through town and never earned a single speeding strike")
        money_delta = rig.ctx.profile.money - 5000.0
        if money_delta != 0.0:
            findings.append(
                f"money moved {money_delta:+,.0f} mid-trip; strikes are supposed to be "
                "settled at delivery"
            )
        return _outcome(
            "floor_it_through_town",
            rig,
            findings,
            f"{d.speeding_strikes} strikes, spoken totals match the ledger",
        )
    finally:
        rig.close()


@scenario(
    "hairpin_at_70_no_assists",
    "Take a 25-mph hairpin at 70 with every assist off; does anything push back?",
)
def _hairpin():
    rig = Rig()
    findings: list[str] = []
    try:
        d = rig.d
        rig.ctx.settings.curve_speed_assist = False
        rig.ctx.settings.route_transition_assist = False
        start = 12.0
        d.trip.position_mi = start
        d.trip.curves = [_fabricated_curve(start + 1.0)]
        rig.prepare(speed_mph=70.0)
        damage_before = d.truck.damage_pct
        rig.held.add(rig.pygame.K_UP)
        rig.step(6000, until=lambda: d.trip.position_mi >= start + 2.0)
        if rig.said("too fast, drifting to the outside") == 0:
            findings.append("no drifting-outside warning through a hairpin taken 45 over")
        min_speed = d.truck.speed_mph
        damage_delta = d.truck.damage_pct - damage_before
        if damage_delta == 0.0:
            findings.append(
                "blew a 25-advisory hairpin at 70 (assists and lane drift off): zero damage, "
                "no crash, no spoken consequence beyond the warning -- the bend cannot "
                "hurt you"
            )
        return _outcome(
            "hairpin_at_70_no_assists",
            rig,
            findings,
            f"hairpin punished the overspeed (damage +{damage_delta:.0f}, min {min_speed:.0f} mph)",
        )
    finally:
        rig.close()


@scenario(
    "reverse_down_the_route",
    "Engage reverse and back down the interstate; position must clamp, someone should object.",
)
def _reverse_route():
    rig = Rig()
    findings: list[str] = []
    try:
        d = rig.d
        d.trip.position_mi = 1.0
        rig.prepare(speed_mph=0.0)
        # The deliberate direction gesture: a FRESH brake press at a standstill,
        # held through the confirmation beat.
        rig.step(5)  # settle with no keys so the press below is a rising edge
        rig.held.add(rig.pygame.K_DOWN)
        rig.step(60)  # > DIRECTION_CHANGE_HOLD_S at DT
        if not d.truck.transmission.in_reverse:
            findings.append("standstill reverse gesture never engaged reverse")
            return _outcome("reverse_down_the_route", rig, findings, "")
        max_reverse_mph = 0.0
        backed_mi = 0.0
        for _ in range(7200):
            d.update(DT)
            max_reverse_mph = max(max_reverse_mph, d.truck.speed_mph)
            backed_mi = 1.0 - d.trip.position_mi
            rig.check_invariants()
            if d.trip.position_mi <= 0.0:
                break
        rig.step(300)  # keep backing against the route origin
        if max_reverse_mph > 11.0:
            findings.append(f"reverse reached {max_reverse_mph:.1f} mph (cap should be ~10)")
        if d.trip.position_mi < 0.0:
            findings.append("backed to a negative route position")
        wrongway = [
            line
            for line in rig.transcript
            if "wrong" in line.lower()
            or "backward" in line.lower()
            or "back the way" in line.lower()
        ]
        if backed_mi >= 0.5 and not wrongway:
            findings.append(
                f"backed {backed_mi:.1f} miles down the interstate (to route mile "
                f"{d.trip.position_mi:.2f}) with no wrong-way or off-route feedback of any "
                "kind after the initial 'Reverse selected' -- a blind player has no way to "
                "know the trip is unwinding"
            )
        if d.hos.driving_min <= 0.0:
            findings.append("reversing at 9 mph never counted as HOS driving time")
        return _outcome(
            "reverse_down_the_route",
            rig,
            findings,
            f"clamped at mile {d.trip.position_mi:.2f}, reverse capped {max_reverse_mph:.1f} mph",
        )
    finally:
        rig.close()


@scenario(
    "slam_reverse_at_speed",
    "Manual box: grab reverse at 60 mph; a real driveline would grenade.",
)
def _slam_reverse():
    from freight_fate.sim.transmission import REVERSE

    rig = Rig(automatic=False)
    findings: list[str] = []
    try:
        d = rig.d
        d.trip.position_mi = 12.0
        rig.prepare(speed_mph=60.0, gear=10)
        rig.held.add(rig.pygame.K_LSHIFT)  # clutch in
        rig.step(8)
        result = d.truck.transmission.request_gear(REVERSE)
        rig.held.discard(rig.pygame.K_LSHIFT)
        if result.ok:
            findings.append(
                "reverse engaged at 60 mph forward with only the clutch pressed: no speed "
                "guard, no grind, no driveline damage model"
            )
        rig.step(900)
        if d.truck.transmission.in_reverse and d.truck.velocity_mps > 1.0:
            findings.append(
                f"rolling forward at {d.truck.speed_mph:.0f} mph in reverse gear; over-rev "
                f"wear is the only consequence (engine wear {d.truck.engine_wear_pct:.1f}%)"
            )
        redline = rig.lines_with("Redline") + rig.lines_with("taking damage")
        if redline and d.truck.damage_pct == 0.0:
            findings.append(
                f"redline warning says 'taking damage, now {d.truck.damage_pct:.0f} percent' "
                "but over-rev only raises engine WEAR -- the spoken damage number never moves"
            )
        return _outcome(
            "slam_reverse_at_speed",
            rig,
            findings,
            "reverse at speed was refused or punished",
        )
    finally:
        rig.close()


@scenario(
    "neutral_coast_mountain",
    "Slam neutral on a 6% descent and ride it; how fast does it get, and does anything object?",
)
def _neutral_coast():
    rig = Rig(automatic=False)
    findings: list[str] = []
    try:
        d = rig.d
        d.trip.position_mi = 12.0
        d.trip.curves = []
        d.trip.grade_at = lambda mile: -0.06
        rig.prepare(speed_mph=55.0, gear=10)
        result = d.truck.transmission.request_gear(0)  # neutral needs no clutch
        if not result.ok:
            findings.append(f"could not select neutral while rolling: {result.message}")
        max_mph = 0.0
        for _ in range(5400):
            d.update(DT)
            max_mph = max(max_mph, d.truck.speed_mph)
            rig.check_invariants()
            if d.trip.position_mi >= d.trip.total_miles - 8.0:
                break
        if max_mph > 110.0:
            findings.append(
                f"neutral coast reached {max_mph:.0f} mph: no mechanical failure, no "
                "runaway-truck event -- the only pushback is the overspeed chime and "
                "settlement fines"
            )
        strike_count = rig.said("Speeding strike")
        # Try to stop it on the service brakes alone from whatever speed remains.
        rig.held.add(rig.pygame.K_DOWN)
        stopped = rig.step(2700, until=lambda: d.truck.speed_mph < 5.0)
        if d.truck.speed_mph >= 5.0:
            findings.append(
                f"service brakes could not stop the neutral runaway ({d.truck.speed_mph:.0f} "
                f"mph after {stopped * DT:.0f}s of full brake, drums {d.truck.brake_temp_c:.0f}C)"
            )
        return _outcome(
            "neutral_coast_mountain",
            rig,
            findings,
            f"peaked {max_mph:.0f} mph, {strike_count} strikes, brakes recovered it",
        )
    finally:
        rig.close()


@scenario(
    "dynamite_parking_brake_at_60",
    "Pull the parking brake valve at 60: flat-spots, no waiting fast-forward, honest speech.",
)
def _dynamite():
    rig = Rig()
    findings: list[str] = []
    try:
        d = rig.d
        d.trip.position_mi = 12.0
        rig.prepare(speed_mph=60.0)
        rig.step(30)
        wear_before = d.truck.tire_wear_pct
        speed = d.truck.speed_mph
        d._toggle_parking_brake()
        if rig.said("dynamited") == 0:
            findings.append("no spoken dynamiting consequence for a parking brake pull at 60")
        from freight_fate.states.driving_core import FLAT_SPOT_WEAR_PCT_PER_MPH

        expected = speed * FLAT_SPOT_WEAR_PCT_PER_MPH
        delta = d.truck.tire_wear_pct - wear_before
        if abs(delta - expected) > 0.3:
            findings.append(
                f"flat-spot wear {delta:.2f}% does not match the model ({expected:.2f}%)"
            )
        if d.trip.waiting:
            findings.append("waiting fast-forward armed by a parking brake pulled at speed")
        rig.step(1200, until=lambda: d.truck.speed_mph < 0.5)
        if d.truck.speed_mph >= 0.5:
            findings.append("spring brakes never brought the truck to a stop")
        if d.trip.waiting:
            findings.append("waiting fast-forward armed itself after the dynamited stop")
        scale = d.trip.effective_time_scale
        if scale > d.trip.time_scale:
            findings.append(f"parked clock is running at {scale:.0f}x without deliberate waiting")
        rig.held.add(rig.pygame.K_UP)
        rig.step(90)
        rig.held.discard(rig.pygame.K_UP)
        if rig.said("Parking brake set") == 0:
            findings.append("throttle against the set parking brake drew no spoken lockout cue")
        return _outcome(
            "dynamite_parking_brake_at_60",
            rig,
            findings,
            f"flat-spotted {delta:.1f}% tread, stopped, no fast-forward, lockout cue spoken",
        )
    finally:
        rig.close()


@scenario(
    "redline_damage_readout",
    "Force a road-driven over-rev; the warning quotes a damage number that must be honest.",
)
def _redline():
    rig = Rig(automatic=False)
    findings: list[str] = []
    try:
        d = rig.d
        t = d.truck
        d.trip.position_mi = 12.0
        d.trip.curves = []
        d.trip.grade_at = lambda mile: -0.10
        rig.prepare(speed_mph=50.0)
        # Pick the tallest gear that is already past damaging revs at this speed.
        gear = next((g for g in range(10, 0, -1) if t.coupled_rpm(g) > t.specs.max_rpm * 1.06), 3)
        t.transmission.gear = gear
        wear_before = t.engine_wear_pct
        rig.step(600, until=lambda: rig.said("Redline") + rig.said("redline") > 0)
        redline_lines = rig.lines_with("redline") + rig.lines_with("Redline")
        if not redline_lines:
            findings.append("sustained over-rev never drew a spoken redline warning")
            return _outcome("redline_damage_readout", rig, findings, "")
        m = re.search(r"damage, now (\d+) percent|Damage (\d+) percent", redline_lines[0])
        spoken_damage = float(m.group(1) or m.group(2)) if m else None
        wear_gained = t.engine_wear_pct - wear_before
        if spoken_damage is not None and abs(spoken_damage - t.damage_pct) > 1.0:
            findings.append(
                f"redline warning spoke {spoken_damage:.0f}% damage but damage is "
                f"{t.damage_pct:.0f}%"
            )
        if wear_gained > 0.1 and t.damage_pct == 0.0 and "damage" in redline_lines[0].lower():
            findings.append(
                'redline warning says the engine is "taking damage, now 0 percent" while the '
                f"harm actually lands on engine WEAR (+{wear_gained:.1f}%) -- the spoken "
                "number will read 0 forever, telling a blind player the abuse is free"
            )
        return _outcome(
            "redline_damage_readout",
            rig,
            findings,
            "redline warning quotes a number that tracks the real harm",
        )
    finally:
        rig.close()

File: crates/freight-fate/src/playtest/break_scenarios/driving_physics.rs
//! Driving/physics abuse: floor it, reverse it, dynamite it (port of
//! `tools/playtest_break_scenarios/driving_physics.py`).
//!
//! Every scenario here puts the truck in a position no sane driver would
//! choose and checks whether the physics, the ledgers, and the spoken cues
//! stay honest about the consequences.

use crate::playtest::breaker::{fabricated_curve, outcome, Outcome, Rig, RigOptions, DT};
use crate::states::base::Key;
use crate::states::driving_core::{hos_of, FLAT_SPOT_WEAR_PCT_PER_MPH};

/// Flooring it must produce a real pull-over or nothing at all.
///
/// This scenario used to check a "speeding strike" ledger: hold nine over for
/// six seconds anywhere on the route and the drive banked a charge, billed at
/// the dock. That was a fine from an officer who was never there, and it is
/// gone. What replaces it is the harder question -- when the road DOES charge
/// you, was there a trooper, and did the player hear them?
pub fn floor_it_through_town() -> Outcome {
    let mut rig = Rig::new(RigOptions::default());
    let mut findings: Vec<String> = Vec::new();
    let money_before = rig.app.ctx.profile.as_ref().map(|p| p.money).unwrap_or(0.0);
    rig.prepare(30.0, None);
    rig.hold(Key::Up);
    let target = 25.0f64.min(rig.drive.trip.total_miles() - 8.0);
    rig.step(
        9000,
        DT,
        Some(&move |rig: &Rig| rig.drive.trip.position_mi >= target),
    );

    // Nothing may bank a charge behind the player's back any more.
    for ghost in ["Speeding strike", "due at delivery", "dollar maximum"] {
        if !rig.lines_with(ghost).is_empty() {
            findings.push(format!(
                "the removed silent speeding charge still speaks: {ghost:?}"
            ));
        }
    }
    let money_now = rig.app.ctx.profile.as_ref().map(|p| p.money).unwrap_or(0.0);
    let money_delta = money_now - money_before;
    let tickets = rig.drive.speeding_tickets;
    let fines = rig.drive.ticket_fines_paid;
    if money_delta < 0.0 && tickets == 0 {
        findings.push(format!(
            "money fell {:.0} with no traffic stop on the record: speeding nobody saw is \
             supposed to cost nothing",
            -money_delta
        ));
    }
    // A ticket is real money, so it has to have had a real officer behind it
    // -- and the player has to have heard them coming.
    if tickets != 0 {
        if rig.lines_with("Lights and siren behind you").is_empty() {
            findings.push(format!(
                "{tickets} ticket(s) written with no lights-and-siren call: the player was \
                 charged by an officer they never heard"
            ));
        }
        if (money_delta + fines).abs() > 0.01 {
            findings.push(format!(
                "ticket ledger mismatch: money moved {money_delta:+.0} but tickets total {fines:.0}"
            ));
        }
    }
    let every_post_heard = rig
        .drive
        .trip
        .posts
        .iter()
        .filter(|post| post.declined || !post.staffed)
        .all(|post| post.announced);
    if !every_post_heard {
        findings.push("a post acted on the driver before it had made a sound".to_string());
    }
    let note =
        format!("{tickets} traffic stop(s), {fines:.0} dollars, every one of them audible first");
    outcome("floor_it_through_town", &rig, findings, &note)
}

pub fn hairpin_at_70_no_assists() -> Outcome {
    let mut rig = Rig::new(RigOptions::default());
    let mut findings: Vec<String> = Vec::new();
    rig.app.ctx.settings.curve_speed_assist = false;
    rig.app.ctx.settings.route_transition_assist = false;
    let start = 12.0;
    rig.drive.trip.position_mi = start;
    rig.drive.trip.curves = vec![fabricated_curve(start + 1.0, 25, 'R')];
    rig.prepare(70.0, None);
    let damage_before = rig.drive.truck().damage_pct;
    rig.hold(Key::Up);
    rig.step(
        6000,
        DT,
        Some(&move |rig: &Rig| rig.drive.trip.position_mi >= start + 2.0),
    );
    if rig.said("too fast, drifting to the outside") == 0 {
        findings.push("no drifting-outside warning through a hairpin taken 45 over".to_string());
    }
    let min_speed = rig.drive.truck().speed_mph();
    let damage_delta = rig.drive.truck().damage_pct - damage_before;
    if damage_delta == 0.0 {
        findings.push(
            "blew a 25-advisory hairpin at 70 (assists and lane drift off): zero damage, no \
             crash, no spoken consequence beyond the warning -- the bend cannot hurt you"
                .to_string(),
        );
    }
    let note = format!(
        "hairpin punished the overspeed (damage +{damage_delta:.0}, min {min_speed:.0} mph)"
    );
    outcome("hairpin_at_70_no_assists", &rig, findings, &note)
}

pub fn reverse_down_the_route() -> Outcome {
    let mut rig = Rig::new(RigOptions::default());
    let mut findings: Vec<String> = Vec::new();
    rig.drive.trip.position_mi = 1.0;
    rig.prepare(0.0, None);
    // The deliberate direction gesture: a FRESH brake press at a standstill,
    // held through the confirmation beat.
    rig.run_frames(5); // settle with no keys so the press below is a rising edge
    rig.hold(Key::Down);
    rig.run_frames(60); // > DIRECTION_CHANGE_HOLD_S at DT
    if !rig.drive.truck().transmission.in_reverse() {
        findings.push("standstill reverse gesture never engaged reverse".to_string());
        return outcome("reverse_down_the_route", &rig, findings, "");
    }
    let mut max_reverse_mph = 0.0f64;
    let mut backed_mi = 0.0;
    for _ in 0..7200 {
        rig.drive.update_frame(&mut rig.app.ctx, DT);
        rig.app.ctx.run_deferred();
        max_reverse_mph = max_reverse_mph.max(rig.drive.truck().speed_mph());
        backed_mi = 1.0 - rig.drive.trip.position_mi;
        rig.check_invariants();
        if rig.drive.trip.position_mi <= 0.0 {
            break;
        }
    }
    rig.run_frames(300); // keep backing against the route origin
    if max_reverse_mph > 11.0 {
        findings.push(format!(
            "reverse reached {max_reverse_mph:.1} mph (cap should be ~10)"
        ));
    }
    if rig.drive.trip.position_mi < 0.0 {
        findings.push("backed to a negative route position".to_string());
    }
    let wrongway = rig.transcript().into_iter().any(|line| {
        let lower = line.to_lowercase();
        lower.contains("wrong") || lower.contains("backward") || lower.contains("back the way")
    });
    if backed_mi >= 0.5 && !wrongway {
        findings.push(format!(
            "backed {backed_mi:.1} miles down the interstate (to route mile {:.2}) with no \
             wrong-way or off-route feedback of any kind after the reverse beep started -- a \
             blind player has no way to know the trip is unwinding",
            rig.drive.trip.position_mi
        ));
    }
    if hos_of(&rig.app.ctx).driving_min <= 0.0 {
        findings.push("reversing at 9 mph never counted as HOS driving time".to_string());
    }
    let note = format!(
        "clamped at mile {:.2}, reverse capped {max_reverse_mph:.1} mph",
        rig.drive.trip.position_mi
    );
    outcome("reverse_down_the_route", &rig, findings, &note)
}

pub fn dynamite_parking_brake_at_60() -> Outcome {
    let mut rig = Rig::new(RigOptions::default());
    let mut findings: Vec<String> = Vec::new();
    rig.drive.trip.position_mi = 12.0;
    rig.prepare(60.0, None);
    rig.run_frames(30);
    let wear_before = rig.drive.truck().tire_wear_pct;
    let speed = rig.drive.truck().speed_mph();
    rig.drive.toggle_parking_brake(&mut rig.app.ctx);
    rig.app.ctx.run_deferred();
    if rig.said("dynamited") == 0 {
        findings
            .push("no spoken dynamiting consequence for a parking brake pull at 60".to_string());
    }
    let expected = speed * FLAT_SPOT_WEAR_PCT_PER_MPH;
    let delta = rig.drive.truck().tire_wear_pct - wear_before;
    if (delta - expected).abs() > 0.3 {
        findings.push(format!(
            "flat-spot wear {delta:.2}% does not match the model ({expected:.2}%)"
        ));
    }
    if rig.drive.trip.waiting {
        findings.push("waiting fast-forward armed by a parking brake pulled at speed".to_string());
    }
    rig.step(
        1200,
        DT,
        Some(&|rig: &Rig| rig.drive.truck().speed_mph() < 0.5),
    );
    if rig.drive.truck().speed_mph() >= 0.5 {
        findings.push("spring brakes never brought the truck to a stop".to_string());
    }
    if rig.drive.trip.waiting {
        findings.push("waiting fast-forward armed itself after the dynamited stop".to_string());
    }
    let scale = rig.drive.trip.effective_time_scale();
    if scale > rig.drive.trip.time_scale {
        findings.push(format!(
            "parked clock is running at {scale:.0}x without deliberate waiting"
        ));
    }
    rig.hold(Key::Up);
    rig.run_frames(90);
    rig.release(Key::Up);
    if rig.said("Parking brake set") == 0 {
        findings
            .push("throttle against the set parking brake drew no spoken lockout cue".to_string());
    }
    let note =
        format!("flat-spotted {delta:.1}% tread, stopped, no fast-forward, lockout cue spoken");
    outcome("dynamite_parking_brake_at_60", &rig, findings, &note)
}
